pkg/ebpf/tracee.go

```diff
@@ -950,49 +1002,28 @@ func (t *Tracee) populateBPFMaps() error {
-	binary.LittleEndian.PutUint32(configVal[8:12], t.getFiltersConfig())
-	binary.LittleEndian.PutUint32(configVal[12:16], uint32(t.containers.GetDefaultCgroupHierarchyID()))
-	binary.LittleEndian.PutUint64(configVal[16:24], t.config.Filter.UIDFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[24:32], t.config.Filter.UIDFilter.Minimum())
-	binary.LittleEndian.PutUint64(configVal[32:40], t.config.Filter.PIDFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[40:48], t.config.Filter.PIDFilter.Minimum())
-	binary.LittleEndian.PutUint64(configVal[48:56], t.config.Filter.MntNSFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[56:64], t.config.Filter.MntNSFilter.Minimum())
-	binary.LittleEndian.PutUint64(configVal[64:72], t.config.Filter.PidNSFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[72:80], t.config.Filter.PidNSFilter.Minimum())
-	// Next 128 bytes (1024 bits) are used for events_to_submit configuration
-	// Set according to events chosen by the user
-	for id, e := range t.events {
-		if id >= 1024 {
-			// we support up to 1024 events shared with bpf code
-			continue
-		}
-		if e.submit {
-			index := id / 8
-			offset := id % 8
-			configVal[80+index] = configVal[80+index] | (1 << offset)
-		}
-	}
+	configVal := t.computeConfigValues()
 	if err = bpfConfigMap.Update(unsafe.Pointer(&cZero), unsafe.Pointer(&configVal[0])); err != nil {
 		return err
 	}
 
-	errmap := make(map[string]error, 0)
-	errmap[UIDFilterMap] = t.config.Filter.UIDFilter.InitBPF(t.bpfModule)
-	errmap[PIDFilterMap] = t.config.Filter.PIDFilter.InitBPF(t.bpfModule)
-	errmap[MntNSFilterMap] = t.config.Filter.MntNSFilter.InitBPF(t.bpfModule)
-	errmap[PidNSFilterMap] = t.config.Filter.PidNSFilter.InitBPF(t.bpfModule)
-	errmap[UTSFilterMap] = t.config.Filter.UTSFilter.InitBPF(t.bpfModule)
-	errmap[CommFilterMap] = t.config.Filter.CommFilter.InitBPF(t.bpfModule)
-	errmap[ContIdFilter] = t.config.Filter.ContIDFilter.InitBPF(t.bpfModule, t.containers)
-	errmap[BinaryFilterMap] = t.config.Filter.BinaryFilter.InitBPF(t.bpfModule)
+	for filterScope := range t.config.FilterScopes.Map() {
+		scopeID := uint(filterScope.ID)
+		errMap := make(map[string]error, 0)
 
-	for k, v := range errmap {
-		if v != nil {
-			return fmt.Errorf("error setting %v filter: %v", k, v)
+		errMap[UIDFilterMap] = filterScope.UIDFilter.InitBPF(t.bpfModule, scopeID)
+		errMap[PIDFilterMap] = filterScope.PIDFilter.InitBPF(t.bpfModule, scopeID)
+		errMap[MntNSFilterMap] = filterScope.MntNSFilter.InitBPF(t.bpfModule, scopeID)
+		errMap[PidNSFilterMap] = filterScope.PidNSFilter.InitBPF(t.bpfModule, scopeID)
+		errMap[UTSFilterMap] = filterScope.UTSFilter.InitBPF(t.bpfModule, scopeID)
+		errMap[CommFilterMap] = filterScope.CommFilter.InitBPF(t.bpfModule, scopeID)
+		errMap[ContIdFilter] = filterScope.ContIDFilter.InitBPF(t.bpfModule, t.containers, scopeID)
+		errMap[BinaryFilterMap] = filterScope.BinaryFilter.InitBPF(t.bpfModule, scopeID)
```

Comment:
I think we should avoid setting the BPF maps for every scope (get value + set value operations).
Instead, the bitmaps can be calculated in userspace, then initialize the maps once.

Reply:
I agree. Currently all the logic for updating filter maps is inside each filter type; depending on the type the logic is different: https://github.com/aquasecurity/tracee/issues/2589. For now, I suppose changing `InitBPF` to `UpdateBPF` would make it clearer.